File: day2/day2.py
def get_invalid_ids(ranges):
    invalids = []
    for range_string in ranges:
        start, end = range_string.split("-")
        start = int(start)
        end = int(end)

        # Anything with a leading 0 is an invalid id
        # Any number comprised of a repeating sequence of digits e.g: 4567845678, 55, 111, 333333
        for num in range(start, end+1):
            num_str = str(num)
            
            # starts with 0
            if num_str.startswith("0"):
                invalids.append(num)
                continue

            # check if comprised only of repeating digits
            double_num_str = num_str + num_str # double the string so 789789 becomes 789789789789
            double_num_str = double_num_str[1:-1] # remove first and last digit

            # if the original string is still in the doubled string after removing first and last digits, its comprised of only repeating digit sequence
            if num_str in double_num_str:
                invalids.append(num)

    return invalids

def main():
    with open("../input/day2_input") as f:
        input_string = f.read()
        ranges = input_string.split(",")
        print(ranges)
        invalid_ids = get_invalid_ids(ranges)
        answer = sum(invalid_ids)
        print(answer)

if __name__ == "__main__":
    main()

File: day2/day2.cpp
#include <sstream>
#include <fstream>
#include <string>
#include <vector>
#include <iostream>

using namespace std;

vector<long long> count_invalid_ids(const vector<string> &ranges)
{
    vector<long long> invalid_ids;

    for (const string &range : ranges)
    {
        long long start, end;
        char dash;

        stringstream ss(range);

        ss >> start >> dash >> end;

        // cout << range << endl;
        // cout << start << "-" << end << endl;

        while (start <= end)
        {
            string num_string = to_string(start);
            char first = num_string[0];
            if (first == '0')
            {
                invalid_ids.push_back(start);
                start++;
                continue;
            }

            string double_string = num_string + num_string;
            double_string.erase(0, 1); // remove first char
            double_string.pop_back();  // remove last char

            if (double_string.find(num_string) != string::npos)
            {
                invalid_ids.push_back(start);
                start++;
                continue;
            }
            start++;
        }
    }
    return invalid_ids;
}

int main()
{
    ifstream file("../input/day2_input");
    stringstream buffer;
    vector<string> ranges;

    buffer << file.rdbuf();
    string input_string = buffer.str();

    stringstream input_stream(input_string);
    string range;

    // Separate each range by comma and add to range vector
    while (getline(input_stream, range, ','))
    {
        ranges.push_back(range);
    }

    file.close();

    vector<long long> invalid_ids = count_invalid_ids(ranges);

    long long sum = 0;

    for (long long num : invalid_ids)
    {
        sum += num;
    }

    cout << "Answer: " << sum << endl;

    return 0;
}